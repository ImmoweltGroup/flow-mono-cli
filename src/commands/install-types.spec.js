// @flow

jest.mock('jest-worker');

jest.mock('./../lib/paths.js');
jest.mock('./../lib/logger.js');
jest.mock('./../lib/dependency.js');
jest.mock('./../lib/flowTyped.js');
jest.mock('./../lib/exec.js');
console.error = jest.fn();

const {default: Worker} = require('jest-worker');

const path: any = require('./../lib/paths.js');
const dependency: any = require('./../lib/dependency.js');
const flowTyped: any = require('./../lib/flowTyped.js');
const logger: any = require('./../lib/logger.js');
const exec: any = require('./../lib/exec.js');

const installFlowTypes = require('./install-types.js');

describe('install-types', () => {
  beforeAll(() => {
    Worker.mockImplementation(() => ({
      asyncWithRetries: exec.asyncWithRetries,
      end: () => {}
    }));
  });

  afterEach(() => {
    // $FlowFixMe: Ignore errors since the jest type-def is out of date.
    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it('should export an function', () => {
    expect(typeof installFlowTypes).toBe('function');
  });

  it('should update the flow-typed cache and afterwards install types in all packages', async () => {
    path.resolveMonoRepoRootPath.mockReturnValue('/foo');
    path.resolveMonoRepoPackagePaths.mockReturnValue(['/foo/bar', '/foo/baz']);
    flowTyped.parseArgs.mockReturnValue(['--overwrite']);

    await installFlowTypes();

    expect(exec.asyncWithRetries.mock.calls).toMatchSnapshot();
  });

  it('should fatally exit the process if something went wrong during the installation', async () => {
    path.resolveMonoRepoRootPath.mockReturnValue('/foo');
    path.resolveMonoRepoPackagePaths.mockReturnValue(['/foo/bar', '/foo/baz']);
    flowTyped.parseArgs.mockReturnValue(['--overwrite']);
    dependency.readPackageJson.mockReturnValue({name: 'myFooPackageName'});
    exec.asyncWithRetries.mockReturnValueOnce(null).mockReturnValueOnce(Promise.reject(new Error('Foo')));

    await installFlowTypes();

    expect(logger.error.mock.calls).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
    expect(console.error.mock.calls[0][0].message).toBe('Foo');
  });
});
