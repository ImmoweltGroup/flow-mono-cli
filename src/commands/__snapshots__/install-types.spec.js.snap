// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`install-types should fatally exit the process if something went wrong during the installation 1`] = `
Array [
  Array [
    "Failed installing \\"flow-typed\\" definitions in package \\"myFooPackageName\\"",
    "Foo",
  ],
]
`;

exports[`install-types should update the flow-typed cache and afterwards install types in all packages 1`] = `
Array [
  Array [
    "flow-typed",
    Array [
      "install",
      "--overwrite",
    ],
    Object {
      "cwd": "/foo/bar",
      "localDir": "/foo/bar",
      "preferLocal": true,
    },
  ],
  Array [
    "flow-typed",
    Array [
      "install",
      "--overwrite",
    ],
    Object {
      "cwd": "/foo/baz",
      "localDir": "/foo/baz",
      "preferLocal": true,
    },
  ],
]
`;
